✨ [Story interactive] Disclaimer with url

mszylkowski opened the pull request:
Closes https://github.com/ampproject/amphtml/issues/31447

Run `AMP.toggleExperiment('amp-story-interactive-disclaimer')` to enable the experiment.

Adds a disclaimer icon that expands to a bubble.

The disclaimer icon is tabbable when inactive, and the bubble link and close button are tabbable when active.

![image](https://user-images.githubusercontent.com/22420856/101798567-323b6680-3ad9-11eb-8704-0db874a22894.png)

![image](https://user-images.githubusercontent.com/22420856/101798593-3c5d6500-3ad9-11eb-8dd0-aff32c15684a.png)

![image](https://user-images.githubusercontent.com/22420856/101799099-bd1c6100-3ad9-11eb-9215-7a4d9e319ebe.png)

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive-abstract.js:
```diff
@@ -363,6 +368,14 @@ export class AmpStoryInteractive extends AMP.BaseElement {
+    ) {
+      // Needs to be called after buildCallback to measure properly.
+      this.disclaimerEl_ = buildInteractiveDisclaimer(this);
+      this.rootEl_.appendChild(this.disclaimerEl_);
```
a11y ux:
On a screen reader this would be read after reading all of the quiz content.
Using `.prepend` can help ensure a screen reader will read this before the quiz content.

You may need to add `z-index: 1;` to the disclaimer CSS since this will change the layout order.

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -17,6 +17,7 @@
 /* 1em = 16px for interactive components, unless overriden by font-size */
 .i-amphtml-story-interactive-container {
   font-family: 'Poppins', sans-serif !important;
+  --i-amphtml-story-interactive-material-easing: cubic-bezier(0.4, 0, 0.2, 1);
```
Since this is in the context of our polls animation UX we can replace all instances of this with 
`--i-amphtml-interactive-ease-out-curve`.

mszylkowski replied:
Removing the animations for now but will do!

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -110,3 +111,144 @@
+.i-amphtml-story-interactive-disclaimer {
+  --i-amphtml-story-interactive-disclaimer-animation-duration: 0.15s !important;
+  position: absolute !important;
+  right: 0.25em !important;
+  top: 0.25em !important;
```
On polls without a prompt the border radius aligns:
<img width="426" alt="Screen Shot 2020-12-11 at 11 30 48 AM" src="https://user-images.githubusercontent.com/3860311/101929382-c45c7100-3ba4-11eb-9eb2-35a0dc3c018e.png">

`.5em` in the cases where the prompt is a sibling looks more consistent:
<img width="428" alt="Screen Shot 2020-12-11 at 11 35 40 AM" src="https://user-images.githubusercontent.com/3860311/101929609-1604fb80-3ba5-11eb-8ecc-624e09740688.png">

mszylkowski replied:
According to mocks the icon will be outside the container so that it doesn't block the options, but this adds more complexity in the code so left it out in this PR
![image](https://user-images.githubusercontent.com/22420856/102100200-27dddd00-3df7-11eb-9fbd-7f1297ca7985.png)

processprocess commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,168 @@
+        urlEl.textContent = backendSpecs[0];
+        backendSpecs[1].learnMoreUrl
+          ? (linkEl.href = backendSpecs[1].learnMoreUrl)
+          : linkEl.remove();
```
If the link is not present it's effecting the scale in the closed state:
<img width="64" alt="Screen Shot 2020-12-11 at 11 40 50 AM" src="https://user-images.githubusercontent.com/3860311/101930173-ce32a400-3ba5-11eb-9bfb-9cbf4599680e.png">

mszylkowski replied:
Left 2 `TODO`s on the CSS rules saying that the scale values depend on the size of the bubble. I have that working [here](https://github.com/ampproject/amphtml/pull/31534/files#diff-aaf53ed659819e653d5c15b8146e40e2e102886ad3375cc0cfcdb963b439bc0eR163) but it was also left out of this PR for simplicity.

processprocess commented:
The work on the transform and scale is great but we're sacrificing dom structure, semantics and maintainability. 
As a result this code is brittle.

Adding more lines to the disclaimer breaks the transform.
An accessible experience requires a re-write. We can't nest buttons.

The DOM structure in this PR was informed by the animation. Structure and semantics always come first.

Please revise this PR without animation and with the dom structure as simple and readable as possible.
This can be as simple as the "!" being a toggle which displays a sibling with the disclaimer content.

mszylkowski commented:
True. In terms of a11y there was an even bigger problem: off-screen links and buttons were being focused when they shouldn't, so the browser would scroll off-screen. I'll fix that with the dom fixes as well, and remove the animations for now as well.

mszylkowski commented:
Update: fixed tabbing and changed the tap target to a separate element instead of catching click events on the container. Removed animations and fixed tabbing by setting `display: none` on the alert when it's active, and on the bubble when it's not active. Also when interactives are not on screen, the disclaimer alert also will be `display: none` to prevent tabbing to it off-screen.

processprocess commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,162 @@
+function buildDisclaimerLayout(element) {
+  const html = htmlFor(element);
+  return html`<div class="i-amphtml-story-interactive-disclaimer">
+    <button class="i-amphtml-story-interactive-disclaimer-alert">
```
Adding `aria-label` to buttons without text will help announce what the button is when using a screen reader. 
This could be:
`aria-label="open disclaimer"` 
i18n could be added in a follow up.

gmajoulet replied:
+1 to the aria-label, and we should do i18n in this PR. Philip was right and this PR is doing too much, even without the animation. A first PR with a simple toggle and without any extra customization would have been a good start, to get a11y/i18n right.

processprocess commented:
Clicking `!` by tabbing to it and pressing enter will trigger navigation. Filed this bug in #31611. The bug is unrelated to the code in this PR. We can stub-out that method to test a11y of this. Just want to make a note of that here :)

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -110,3 +110,93 @@
+}
+
+.i-amphtml-story-interactive-disclaimer-alert {
+  width: 1.75em !important;
+  height: 1.75em !important;
```
buttons have a system font by default that breaks relative em sizing.
`font-size: inherit;` can fix this.
<img width="444" alt="Screen Shot 2020-12-15 at 2 30 29 PM" src="https://user-images.githubusercontent.com/3860311/102263296-1ae9e800-3ee2-11eb-89eb-11f12cf37439.png">

gmajoulet commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,162 @@
+    </button>
+    <div class="i-amphtml-story-interactive-disclaimer-bubble">
+      <div>
+        <span>Your response will be sent to </span>
```
These strings need to be translated

gmajoulet commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,162 @@
+ */
+function buildDisclaimerLayout(element) {
+  const html = htmlFor(element);
+  return html`<div class="i-amphtml-story-interactive-disclaimer">
```
How do you make this not navigate to the next story page on click? I thought we had to add a `role="button"` on the tap target.

mszylkowski replied:
With `pointer-events` on the bubble (open state) and alert (closed state)

processprocess replied:
> With pointer-events on the bubble (open state) and alert (closed state) 

What do you mean?

I thought it was happening here:
https://github.com/ampproject/amphtml/blob/7395ebfebd0bdcfd0deb1c00a54f6d1878662c85/extensions/amp-story/1.0/page-advancement.js#L483

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -110,3 +110,93 @@
+
+/* Disclaimer CSS rules */
+.i-amphtml-story-interactive-disclaimer {
+  --i-amphtml-story-interactive-disclaimer-animation-duration: 0.15s !important;
```
We can remove this for now.

processprocess commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,177 @@
+      class="i-amphtml-story-interactive-disclaimer-alert"
+      aria-label="Open disclaimer"
+    >
+      <div class="i-amphtml-story-interactive-disclaimer-alert-icon"></div>
```
This could be a pseudo element of  `.i-amphtml-story-interactive-disclaimer-alert` since it's decorative.

Then add  `content: '';` and  `display: block; to the styling`
I think you could also remove `pointer-events: none` if you do that too.

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -110,3 +110,92 @@
+  right: 0 !important;
+  cursor: pointer !important;
+  background: url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" fill="#9AA0A6"><rect x="0.787868" y="1.83337" width="1.47851" height="13.2636" rx="0.739255" transform="rotate(-45 0.787868 1.83337)"/><rect x="10.1669" y="0.787868" width="1.47851" height="13.2636" rx="0.739255" transform="rotate(45 10.1669 0.787868)"/></svg>') !important;
+  width: 3em !important;
+  height: 3em !important;
```
Adding `font-size: inherit;` will make this size responsive since a system default is set on buttons.

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive-abstract.js:
```diff
@@ -169,6 +174,9 @@ export class AmpStoryInteractive extends AMP.BaseElement {
     this.rootEl_ = null;
 
+    /** @public {../../../src/service/localizationService} */
+    this.localizationService = Services.localizationForDoc(this.element);
```
This is async and might not be ready when the page loads. You might need to get this in the buildcallback before it's resolved.
<img width="895" alt="Screen Shot 2020-12-17 at 10 59 19 AM" src="https://user-images.githubusercontent.com/3860311/102511678-11ce5780-4057-11eb-9d53-05b6e9ec4a9f.png">

Example:
https://github.com/ampproject/amphtml/blob/master/extensions/amp-story-360/0.1/amp-story-360.js#L402-L407

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -110,3 +110,92 @@
+
+.i-amphtml-story-interactive-disclaimer[active]
+  .i-amphtml-story-interactive-disclaimer-alert {
+  display: none !important;
```
This causes a slight flash on navigation since `CURRENT_PAGE_ID` service is slightly async.
![Dec-17-2020 11-10-26](https://user-images.githubusercontent.com/3860311/102517178-bce20f80-405d-11eb-840e-dbb2d875115a.gif)

We could set tabindex within the `CURRENT_PAGE_ID` listener.

It's not perfect though because we might want to set a tabindex on the options buttons when we work on a11y for that.

This is a tricky one since aria-hidden doesn't stop tabbing to a focusable element.
https://web.dev/aria-hidden-focus/

mszylkowski replied:
Setting the tabindex is much more involved (because we have two buttons and a link that we'd need to change the tabindex) versus just hiding it, so there's a tradeoff. Given that the flash only happens because those two elements are exactly the same, I'd assume that in any other case where the alert doesn't show exactly on the same spot it wouldn't be seen as a flash.

However, we will have to take care of the options when we make them buttons... WDYT if we do `this.element.querySelectorAll("a, button").forEach((e) => e.setAttribute("tabindex", "-1"))` (and `setAttribute("tabindex", "0")` to enable)?

processprocess replied:
The flash is visible when the elements are in different positions.
![Dec-17-2020 12-25-27](https://user-images.githubusercontent.com/3860311/102521530-0da83700-4063-11eb-8b21-34824c10ee8c.gif)

I think that sounds good for the tabindex. If we want to set an order of tabindex in the future we would have to change this but maybe this works for now.

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -110,3 +110,94 @@
+  right: 0 !important;
+  bottom: 0 !important;
+  transform-origin: bottom right !important;
+  pointer-events: all !important;
```
I wonder if we can remove this here and in `i-amphtml-story-interactive-disclaimer-bubble` and put it once on `i-amphtml-story-interactive-disclaimer`.

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -110,3 +110,94 @@
+  display: none !important;
+}
+
+.i-amphtml-story-interactive-disclaimer-alert::after {
```
This could be part of `i-amphtml-story-interactive-disclaimer-alert`.
It's kind a big block but maybe it's better to not have he pseudo element in the dom.

.i-amphtml-story-interactive-disclaimer-alert {
  width: 1.75em !important;
  height: 1.75em !important;
  position: absolute !important;
  right: 0 !important;
  bottom: 0 !important;
  padding: .25em !important;
  font-size: inherit !important;
  transform-origin: bottom right !important;
  background-color: white !important;
  background-image: url('svg code') !important;
  background-repeat: no-repeat !important;
  background-position: center !important;
  background-size: 0.25em !important;
  border: .25em solid var(--interactive-accent-color) !important;
  border-radius: 50% !important;
  box-shadow: 0 .2em .2em 0 #00000040 !important;
  box-sizing: border-box !important;
}

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -110,3 +110,94 @@
+  font-size: 1em !important;
+  background: white;
+  border-radius: 1em;
+  box-shadow: 0px .2em .2em 0px #00000040 !important;
```
`#00000040` could be a top level var since it's used more than once.
It might seem like overkill but colors can get out of hand very fast.

mszylkowski replied:
Also moved the blue link color to a variable.

gmajoulet commented:
Please wait for https://github.com/ampproject/amphtml/pull/31615 to be merged before merging this one, as we don't want to regenerate the localization files JSON migration. Thanks!

mszylkowski commented:
@gmajoulet merged the json translations into here after the other PRs were merged to master

gmajoulet commented on extensions/amp-story-interactive/0.1/amp-story-interactive-abstract.js:
```diff
@@ -788,4 +810,14 @@ export class AmpStoryInteractive extends AMP.BaseElement {
+   * Toggles the tabbable elements (buttons, links, etc) to only reach them when page is active.
+   * @param {boolean} toggle
+   */
+  toggleTabbableElements_(toggle) {
```
The entire page has some accessibility attributes that should ensure that none of its content is tabbable. This method should be a no-op, and if it's not we should file a ticket and solve this problem at a high level.

mszylkowski replied:
When I tab on desktop, it does focus the tabbable elements if I don't fix it from here:
![image](https://user-images.githubusercontent.com/22420856/103236046-ac5d4f00-4911-11eb-8e53-3535b1f36528.png)

Checking now, links on other pages also get focused! This is definitely a higher-level problem but quizzes working inside a shadow-root might not be influenced by whatever logic we write on the story pages to disable tabbing through other pages. Ticket is https://github.com/ampproject/amphtml/issues/31747 for the general problem, but as I said, we might need to implement sth element-wise if the amp-story generalization doesn't apply to shadow-root elements...

processprocess replied:
I wonder if we want to take out `toggleTabbableElements_` and save it for the work in #31747.
Or we could make a TODO to remove it when  #31747 is fixed.

mszylkowski replied:
Added the `TODO`

gmajoulet commented on extensions/amp-story/1.0/_locales/en.json:
```diff
@@ -246,5 +246,9 @@
+  },
+  "89": {
+    "string": "Your response will be sent to ",
+    "description": "Text that explains where the response will be sent on quizzes and polls disclaimers, ending with space"
```
This lacks context for a translator.
They don't know where/how the text will be displayed: should they be very concise? Is it okay to use more words to keep more context?
They don't know what "quizzes and polls disclaimers" are.
This should not end with a space. This is confusing at best, and we don't know if the software they use, or linters, even allow for that.

Another example: "Text displayed to users after clicking a button that reveals a disclaimer, telling them more about where their user data will be stored, after they interact with a poll or quiz."

mszylkowski replied:
I agree the text you suggest is more clear.

I'd say there's something very important about including the space at the end of the translated string in English, which is that "to" as a proposition is a separate word in English but not in Hebrew for instance, where "to" is a prefix of the object (eg: "to him" is just one word, imagine "tohim").

If technically it's a problem, we could add a margin or add it in some other way, but having a space before the entity is not a generic guarantee in all languages. I wonder if it would make sense to introduce a substitution/format to add the Entity in case in other languages it wouldn't conform to being last (eg: `"Your response will be sent to %s"`) That's why having the space as part of the translation seemed better than not. WDYT?

mszylkowski replied:
Well, space seems to be provided for free on inline elements so fighting against that is not a good idea, will follow your suggestion

gmajoulet reviewed (commented):
Some a11y/i18n comments, deferring to @processprocess for the code review and approval

newmuis commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,175 @@
+const BACKENDS = {
+  'webstoriesinteractivity-beta.web.app': {
+    learnMoreUrl: 'https://policies.google.com/terms',
+    entityName: 'AMP',
```
nit: let's make sure it's clear that this is a test page.

newmuis replied:
Sorry, I also mean: let's not name this "AMP", but something like "AMP Interactivity Disclaimer Test Site," to avoid the abuse vector where anyone (including you, or someone who gains access to your Firebase instance) from being able to impersonate AMP more broadly.

newmuis commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,175 @@
+/** @const {!Object<string, !Object>} */
+const BACKENDS = {
+  'webstoriesinteractivity-beta.web.app': {
+    learnMoreUrl: 'https://policies.google.com/terms',
```
nit: can we link this to amp.dev/stories or similar?

processprocess commented on extensions/amp-story-interactive/0.1/amp-story-interactive.css:
```diff
@@ -110,3 +110,83 @@
+.i-amphtml-story-interactive-disclaimer {
+  position: absolute !important;
+  color: black !important;
+  bottom: calc(100% - 1.75em) !important;
```
It looks like this breaks when the quiz is at the top of the page:
<img width="492" alt="Screen Shot 2021-01-05 at 12 56 55 PM" src="https://user-images.githubusercontent.com/3860311/103681811-f53a8680-4f55-11eb-9c37-6bc775e62942.png">

mszylkowski replied:
Good catch! I'd say we should check if the top position of the bubble leaves enough space for the bubble to expand up, and if not, move it down instead.

Filed a bug at #31799 since it would be a lot to include that in this PR

newmuis commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,176 @@
+ */
+
+/** @const {!Object<string, !Object>} */
+const BACKENDS = {
```
Actually, let's move this to its own file and import it here, since this is effectively a feature allowlist/config.

processprocess commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,170 @@
+      class="i-amphtml-story-interactive-disclaimer-alert"
+      aria-label="Open disclaimer"
+    ></button>
+    <div class="i-amphtml-story-interactive-disclaimer-bubble">
```
Per our research:
role=alertdialog could be great to announce this content to screen reader users when this is opened.
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alertdialog_role

It might need to be on a wrapper of the -note and -entity.

processprocess commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,173 @@
+    ></button>
+    <div
+      class="i-amphtml-story-interactive-disclaimer-bubble"
+      role="alertdialog"
```
Per our research:
aria-describedby can help the screen reader announce the content when this is opened.

processprocess commented on extensions/amp-story-interactive/0.1/interactive-disclaimer.js:
```diff
@@ -0,0 +1,179 @@
+      class="i-amphtml-story-interactive-disclaimer-alert"
+      aria-label="Open disclaimer"
+    ></button>
+    <div
+      class="i-amphtml-story-interactive-disclaimer-bubble"
+      role="alertdialog"
```
`aria-describedby` needs to be on the same element as `role="alertdialog"`. It looks like it's on `i-amphtml-story-interactive-disclaimer` right now.
<img width="614" alt="Screen Shot 2021-01-07 at 4 19 45 PM" src="https://user-images.githubusercontent.com/3860311/103946150-4ee0b380-5104-11eb-8d8b-e0974e170c70.png">

processprocess reviewed (approved):
Looks good to me. Thanks Matias!

mszylkowski commented:
@samouri for localized strings owners

samouri reviewed (approved):
Approving for `src/localized-strings.js`
